sophia/src/term/blank_node.rs

```diff
@@ -0,0 +1,287 @@
+    /// [`BLANK_NODE_LABEL`](struct.BLANK_NODE_LABEL.html). This means that it
+    /// must not include the typical leading `_:`.
+    pub fn new<U>(id: U) -> Result<Self>
+    where
+        U: AsRef<str>,
+        TD: From<U>,
+    {
+        if BLANK_NODE_LABEL.is_match(id.as_ref()) {
+            Ok(BlankNode(id.into()))
+        } else {
+            Err(TermError::InvalidBlankNodeId(id.as_ref().to_string()))
+        }
+    }
+
+    /// Return a new blank node with the given identifier.
+    ///
+    /// # Safety
+    ///
+    /// This function requires that `id` is a valid blank node identifier.
+    pub unsafe fn new_unchecked<U>(id: U) -> Self
+    where
+        TD: From<U>,
+    {
+        BlankNode(id.into())
+    }
+
+    /// Copy self while transforming the inner `TermData` with the given
+    /// factory.
+    pub fn copy_with<'a, U, F>(&'a self, mut factory: F) -> BlankNode<U>
+    where
+        U: TermData,
+        F: FnMut(&'a str) -> U,
+    {
+        BlankNode(factory(self.as_ref()))
+    }
+
+    /// Writes the blank node to the `fmt::Write` using the N3 syntax.
+    pub fn write_fmt<W>(&self, w: &mut W) -> fmt::Result
+    where
+        W: fmt::Write,
+    {
+        w.write_str("_:")?;
+        w.write_str(self.as_ref())
+    }
+
+    /// Writes the blank node to the `io::Write` using the N3 syntax.
+    pub fn write_io<W>(&self, w: &mut W) -> io::Result<()>
+    where
+        W: io::Write,
+    {
+        w.write_all(b"_:")?;
+        w.write_all(self.as_ref().as_bytes())
+    }
+
+    /// Return a copy of this blank nodes's underlying identifier.
+    pub fn value(&self) -> String {
+        self.as_ref().to_owned()
+    }
+
+    /// Whether this blank node identifier is compatible with the N3 family
+    /// of syntaxes (N-Triples, Turtle...).
+    ///
+    /// This is already checked by the standard constructor
+    /// [`new()`](struct.BlankNode.html#mehod.new).
+    pub fn is_n3(&self) -> bool {
+        BLANK_NODE_LABEL.is_match(self.as_ref())
+    }
+}
+
+impl<TD> fmt::Display for BlankNode<TD>
+where
+    TD: TermData,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.write_fmt(f)
+    }
+}
+
+impl<TD> AsRef<str> for BlankNode<TD>
+where
+    TD: TermData,
+{
+    fn as_ref(&self) -> &str {
+        self.0.as_ref()
+    }
+}
+
+impl<TD> Deref for BlankNode<TD>
+where
+    TD: TermData,
+{
+    type Target = TD;
+
+    fn deref(&self) -> &TD {
+        &self.0
+    }
+}
+
+impl<T, U> PartialEq<BlankNode<U>> for BlankNode<T>
+where
+    T: TermData,
+    U: TermData,
+{
+    fn eq(&self, other: &BlankNode<U>) -> bool {
+        self.as_ref() == other.as_ref()
+    }
+}
+
+impl<TD> PartialEq<str> for BlankNode<TD>
+where
+    TD: TermData,
+{
+    fn eq(&self, other: &str) -> bool {
+        self.as_ref() == other
+    }
+}
+
+impl<TD> Hash for BlankNode<TD>
+where
+    TD: TermData,
+{
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state)
+    }
+}
```

Comment:
This can actually be replaced by a simple `derive Hash`. Should have been so from the start.